EpiAware/src/EpiLatentModels/models/RandomWalk.jl

```diff
@@ -13,81 +13,78 @@ Z_t = Z_0 + \sigma \sum_{i = 1}^t \epsilon_t
-
-```julia
+```jldoctest RandomWalk; output = false
 using Distributions, Turing, EpiAware
-
-# Create a RandomWalk model
-rw = RandomWalk(init_prior = Normal(2., 1.),
-                                std_prior = HalfNormal(0.1))
+rw = RandomWalk()
+rw
+nothing
+# output
 ```
 
-Then, we can use `generate_latent` to construct a Turing model for a 10 step random walk.
-
-```julia
-# Construct a Turing model
-rw_model = generate_latent(rw, 10)
+```jldoctest RandomWalk; output = false
+mdl = generate_latent(rw, 10)
+mdl()
+nothing
+# output
 ```
 
-Now we can use the `Turing` PPL API to sample underlying parameters and generate the
-unobserved infections.
+```jldoctest RandomWalk; output = false
+rand(mdl)
+nothing
+# output
 
-```julia
-#Sample random parameters from prior
-θ = rand(rw_model)
-#Get random walk sample path as a generated quantities from the model
-Z_t, _ = generated_quantities(rw_model, θ)
 ```
 "
-@kwdef struct RandomWalk{D <: Sampleable, S <: Sampleable} <: AbstractTuringLatentModel
+@kwdef struct RandomWalk{
+    D <: Sampleable, E <: AbstractTuringLatentModel} <:
+              AbstractTuringLatentModel
     init_prior::D = Normal()
-    std_prior::S = HalfNormal(0.25)
+    ϵ_t::E = HierarchicalNormal()
 end
 
 @doc raw"
-Implement the `generate_latent` function for the `RandomWalk` model.
+Generate a latent RW series using accumulate_scan.
 
-## Example usage of `generate_latent` with `RandomWalk` type of latent process model
+# Arguments
 
-```julia
-using Distributions, Turing, EpiAware
+- `latent_model::RandomWalk`: The RandomWalk model.
+- `n::Int`: The length of the RW series.
 
-# Create a RandomWalk model
-rw = RandomWalk(init_prior = Normal(2., 1.),
-                                std_prior = HalfNormal(0.1))
-```
+# Returns
+- `rw::Vector{Float64}`: The generated RW series.
 
-Then, we can use `generate_latent` to construct a Turing model for a 10 step random walk.
-
-```julia
-# Construct a Turing model
-rw_model = generate_latent(rw, 10)
-```
-
-Now we can use the `Turing` PPL API to sample underlying parameters and generate the
-unobserved infections.
-
-```julia
-#Sample random parameters from prior
-θ = rand(rw_model)
-#Get random walk sample path as a generated quantities from the model
-Z_t, _ = generated_quantities(rw_model, θ)
-```
+# Notes
+- `n` must be greater than 0.
 "
 @model function EpiAwareBase.generate_latent(latent_model::RandomWalk, n)
-    σ_RW ~ latent_model.std_prior
+    @assert n>0 "n must be greater than 0"
+
     rw_init ~ latent_model.init_prior
-    ϵ_t ~ filldist(Normal(), n - 1)
-    rw = rw_init .+ vcat(0.0, σ_RW .* cumsum(ϵ_t))
+    @submodel ϵ_t = generate_latent(latent_model.ϵ_t, n - 1)
+
+    rw = accumulate_scan(RWStep(), rw_init, ϵ_t)
+
     return rw
 end
+
+@doc raw"
+The random walk (RW) step function struct
+"
+struct RWStep <: AbstractAccumulationStep end
+
+@doc raw"
+The random walk (RW) step function for use with `accumulate_scan`.
+"
+function (rw::RWStep)(state, ϵ)
```

Comment:
Rather than follow through with the same pattern all the way down to the calculation, couldn't we make a method here like

```julia
function accumulate_scan(step::RWStep, init, ϵ_t)
rw_init .+ vcat(0.0, cumsum(ϵ_t))
end
```

`Base.cumsum` has its own optimisations so grabbing those via dispatch feels very "Julian".

Reply:
Or possibly `Base.accumulate(+, ϵ_t; init = rw_init)`? I think we did a test on this at some point.

Reply:
OK retest:

```julia
using Pkg
Pkg.activate(;temp=true)
using BenchmarkTools
n = 100_000;
ϵ_t = randn(n);
init = randn();

#cumsum version with init
cumsum_init(x, init) = init .+ vcat(zero(init), x)

#accumulate version
cumsum_acc(x, init) = accumulate(+, x; init)

#Check same
cumsum_acc(ϵ_t, init) ≈ cumsum_acc(ϵ_t, init)
# true

@btime cumsum_init(ϵ_t, init);
#  31.750 μs (6 allocations: 1.53 MiB)

@btime cumsum_acc(ϵ_t, init);
#  93.791 μs (4 allocations: 781.33 KiB)
```

The `Base.cumsum` dispatch would be probably quite a bit more efficient.

Reply:
Did a quick check and its about the same.

Reply:
yes agree. I looked at this a bit and found no real difference in performance so went with the most simple (and because in the future if we try and compose steps we will want them as steps) but happy to consider changing it up. I would prefer in its own issue though so we can get a cleaner view of the benchmark implications